hls-test-utils/src/Test/Hls.hs

```diff
@@ -9,38 +9,108 @@ module Test.Hls
 
--- | ingredient: xml runner writes json file of test results (https://github.com/ocharles/tasty-ant-xml/blob/master/Test/Tasty/Runners/AntXML.hs)
---               rerunningTests allow rerun of failed tests (https://github.com/ocharles/tasty-rerun/blob/master/src/Test/Tasty/Ingredients/Rerun.hs)
+-- | Run 'defaultMainWithRerun' with -j1, and silence stderr
 defaultTestRunner :: TestTree -> IO ()
-defaultTestRunner =
-  defaultMainWithIngredients
-    [antXMLRunner, rerunningTests [listingTests, consoleTestReporter]]
+defaultTestRunner = muteStderr . defaultMainWithRerun . adjustOption (const $ NumThreads 1)
 
 gitDiff :: FilePath -> FilePath -> [String]
 gitDiff fRef fNew = ["git", "diff", "--no-index", "--text", "--exit-code", fRef, fNew]
 
 goldenGitDiff :: TestName -> FilePath -> IO ByteString -> TestTree
 goldenGitDiff name = goldenVsStringDiff name gitDiff
 
-testCommand :: String
-testCommand = "test-server"
+runSessionWithServer :: PluginDescriptor IdeState -> FilePath -> Session a -> IO a
+runSessionWithServer plugin = runSessionWithServer' [plugin] def def fullCaps
+
+runSessionWithServerFormatter :: PluginDescriptor IdeState -> String -> FilePath -> Session a -> IO a
+runSessionWithServerFormatter plugin formatter =
+  runSessionWithServer'
+    [plugin]
+    def {formattingProvider = T.pack formatter}
+    def
+    fullCaps
+
+-- | Silence stderr, running an action
+muteStderr :: IO () -> IO ()
+muteStderr action = withTempFile $ \tmp ->
+  bracket (openFile tmp AppendMode) hClose $ \h -> do
+    old <- hDuplicate stderr
+    h `hDuplicateTo'` stderr
+    bracket_ action (hClose old) (old `hDuplicateTo'` stderr)
+
+-- | Host a server, and run a test session on it
+-- Note: cwd will be shifted into @root@ in @Session a@
+runSessionWithServer' ::
+  -- | plugins to load on the server
+  [PluginDescriptor IdeState] ->
+  -- | lsp config for the server
+  Config ->
+  -- | config for the test session
+  SessionConfig ->
+  ClientCapabilities ->
+  FilePath ->
+  Session a ->
+  IO a
+runSessionWithServer' plugin conf sconf caps root s = do
+  (inR, inW) <- createPipe
+  (outR, outW) <- createPipe
+  -- restore cwd after running the session; otherwise the path to test data will be invalid
+  cwd <- getCurrentDirectory
+  let server =
+        Ghcide.defaultMain
+          def
+            { argsHandleIn = pure inR,
+              argsHandleOut = pure outW,
+              argsDefaultHlsConfig = conf,
+              argsIdeOptions = \config sessionLoader ->
+                let ideOptions = (argsIdeOptions def config sessionLoader) {optTesting = IdeTesting True}
+                 in ideOptions {optShakeOptions = (optShakeOptions ideOptions) {shakeThreads = 2}},
+              argsHlsPlugins = pluginDescToIdePlugins $ plugin ++ Ghcide.descriptors
+            }
+  x <- withAsync server $ \_ ->
+    runSessionWithHandles inW outR sconf caps root s
+      `finally` setCurrentDirectory cwd
+  sleep 0.5
```

Comment:
`0.5` is an arbitrary value. Perhaps we could use barrier here. But how do we know that all services are really shut down?

Reply:
Don't use `withAsync`. Instead, start the server using a normal `async` and then, once `runSession...` is over, wait on the server (with a timeout)